File: src/lib/stores/workspace.ts
import { writable, derived } from 'svelte/store';
import { invoke } from '@tauri-apps/api/core';
import type { Workspace } from '../../types';

function createWorkspaceStore() {
    const { subscribe, set, update } = writable<Workspace[]>([]);

    return {
        subscribe,
        loadWorkspaces: async () => {
            try {
                const workspaces = await invoke<Workspace[]>('get_workspaces');
                set(workspaces);
            } catch (error) {
                console.error('Error loading workspaces:', error);
                throw error;
            }
        },
        getWorkspace: async (id: string) => {
            try {
                const workspace = await invoke<Workspace>('get_workspace', { id });
                return workspace;
            } catch (error) {
                console.error('Error getting workspace:', error);
                throw error;
            }
        },
        createWorkspace: async (name: string, description?: string) => {
            try {
                const workspace = await invoke<Workspace>('create_workspace', {
                    name,
                    description,
                });
                update(workspaces => [...workspaces, workspace]);
                return workspace;
            } catch (error) {
                console.error('Error creating workspace:', error);
                throw new Error(error instanceof Error ? error.message : 'Failed to create workspace');
            }
        },
        updateWorkspace: async (id: string, name: string, description?: string) => {
            try {
                const workspace = await invoke<Workspace>('update_workspace', {
                    id,
                    name,
                    description,
                });
                update(workspaces => 
                    workspaces.map(w => w.id === id ? workspace : w)
                );
                return workspace;
            } catch (error) {
                console.error('Error updating workspace:', error);
                throw new Error(error instanceof Error ? error.message : 'Failed to update workspace');
            }
        },
        deleteWorkspace: async (id: string) => {
            try {
                await invoke('delete_workspace', { id });
                update(workspaces => workspaces.filter(w => w.id !== id));
            } catch (error) {
                console.error('Error deleting workspace:', error);
                throw new Error(error instanceof Error ? error.message : 'Failed to delete workspace');
            }
        },
    };
}

export const workspaces = createWorkspaceStore();
export const activeWorkspace = writable<Workspace | null>(null);

// Derived store for the current workspace's variables
export const currentVariables = derived(
    activeWorkspace,
    $workspace => $workspace?.variables || []
); 

File: src-tauri/src/commands/workspace.rs
use crate::storage::{Storage, Workspace};
use std::sync::Mutex;
use tauri::State;

#[tauri::command]
pub async fn create_workspace(
    state: State<'_, Mutex<Storage>>,
    name: String,
    description: Option<String>,
) -> Result<Workspace, String> {
    let mut storage = state.lock().map_err(|e| e.to_string())?;
    storage
        .create_workspace(&name, description.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_workspaces(
    state: State<'_, Mutex<Storage>>,
) -> Result<Vec<Workspace>, String> {
    let mut storage = state.lock().map_err(|e| e.to_string())?;
    storage.get_workspaces().map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_workspace(
    state: State<'_, Mutex<Storage>>,
    id: String,
) -> Result<Option<Workspace>, String> {
    let mut storage = state.lock().map_err(|e| e.to_string())?;
    storage.get_workspace(&id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_workspace(
    state: State<'_, Mutex<Storage>>,
    id: String,
    name: String,
    description: Option<String>,
) -> Result<Workspace, String> {
    let mut storage = state.lock().map_err(|e| e.to_string())?;
    storage
        .update_workspace(&id, &name, description.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_workspace(
    state: State<'_, Mutex<Storage>>,
    id: String,
) -> Result<(), String> {
    let mut storage = state.lock().map_err(|e| e.to_string())?;
    storage.delete_workspace(&id).map_err(|e| e.to_string())
} 
